NPEP: FQDN Selector for Egress, User stories (#134)
kubernetes-sigs/network-policy-api: npep/npep-133.md

# NPEP-133: FQDN Selector for Egress Traffic

* Issue:
  [#133](https://github.com/kubernetes-sigs/network-policy-api/issues/133)
* Status: Provisional

## TLDR

This enhancement proposes adding a new optional selector to specify egress peers
using [Fully Qualified Domain Names](https://www.wikipedia.org/wiki/Fully_qualified_domain_name)
(FQDNs).

## Goals

* Provide a selector to specify egress peers using a Fully Qualified Domain Name
  (for example `kubernetes.io`).

> * Support a restricted set of regex matching capabilities when specifying FQDNs.

Author: Ack, reworded

* Currently only `ALLOW` type rules are proposed.
  * Correctly enforcing DENY rules based on FQDN selectors is difficult as there
    is no guarantee a Network Policy plugin is aware of all IPs backing a FQDN
    policy. 
* Currently only AdminNetworkPolicy is the intended scope for this proposal.

>   * Since Kubernetes NetworkPolicy does not have a FQDN selector, adding this
>     capability to BaselineAdminNetworkPolicy can result in unintended behavior.

Reviewer: ```suggestion
  * Since Kubernetes NetworkPolicy does not have a FQDN selector, adding this
    capability to BaselineAdminNetworkPolicy would not be very useful.
```

Saying that it "can result in unintended behavior" makes it sound like there's some tricky unexpected interaction, but that's not the problem. The problem is just that all of the good user stories that would make use of FQDN BANPs only work if you have FQDN NPs too.

Author: Yeah, good point. Reworded

    For example, if BANP allows traffic to `example.io`, but the namespace admin
    installs a Kubernetes Network Policy, the namespace admin has no way to
    replicate the `example.io` selector using just Kubernetes Network Policies.

## Non-Goals

* This enhancement does not include a FQDN selector for allowing ingress
  traffic.
* This enhancement does not include any L7 matching or filtering capabilities,
  like matching HTTP traffic or URL paths.
  * This selector should not control what DNS records are resolvable from a
    particular workload.
* This enhancement does not provide a mechanism for selecting in-cluster
  endpoints using FQDNs. This is explicitly disallowed by the spec.
  * To select Pods, Nodes, or the API Server, AdminNetworkPolicy has other more

>     specific selectors.

Author: Ack -- I reworded to be more inline with our discussion for Egress CIDRs.

I think leaving it undefined is the right course of action. While for Pods it's pretty trivial to expect providers to do the expected thing (it's the same as putting the Pod's IP directly in). But for something like Services things get tricky since often times providers have already load balanced Service VIPs to a Pod IP before NetPol enforcement.

We can provided better guidance and UX by using more specific selectors for K8s endpoints, so leaving this undefined seems like the best compromise. LMK what you think.

# NPEP-133: FQDN Selector for Egress Traffic

* Issue:
  [#133](https://github.com/kubernetes-sigs/network-policy-api/issues/133)
* Status: Provisional

## TLDR

This enhancement proposes adding a new optional selector to specify egress peers
using [Fully Qualified Domain Names](https://www.wikipedia.org/wiki/Fully_qualified_domain_name)
(FQDNs).

## Goals

* Provide a selector to specify egress peers using a Fully Qualified Domain Name
  (for example `kubernetes.io`).
* Support basic wildcard matching capabilities when specifying FQDNs (for example `*.cloud-provider.io`)
* Currently only `ALLOW` type rules are proposed.
  * Correctly enforcing DENY rules based on FQDN selectors is difficult as there

>     is no guarantee a Network Policy plugin is aware of all IPs backing a FQDN

Author: Ack, good point. I took a pass at fleshing out the argument more clearly. PTAL

# NPEP-133: FQDN Selector for Egress Traffic

* Issue:
  [#133](https://github.com/kubernetes-sigs/network-policy-api/issues/133)
* Status: Provisional

## TLDR

This enhancement proposes adding a new selector to specify egress peers using
Fully Qualified Domain Names (FQDNs). 

> ## Goals

Reviewer 2: As stated below.... Let's be explicit... will this selector be part of ALL current Network Policy API resources?  i.e ANP and BANP?

> ## Goals

Reviewer: Need to define the expected semantics as well; everyone _wants_ the semantics to be "all outbound connections are magically matched against the FQDN rules and it just works", but we know that won't really work.

- Do we support FQDN rules for both Allow and Deny rules?
  - If we support both Allow and Deny rules, what happens if, for example, someone allows connections to `example.com` but denies connections to `example.net` (which has the same IP).
- Do we require that pods make a DNS request before trying to connect, and do we require them to make that DNS request in a particular way (eg, using the resolver pointed to by the kubelet-provided `/etc/resolv.conf`)? Or do we expect the FQDN rule to work correctly even if the pod got the IP address via some means unknown to us?
  - (This is an example of the difference between Allow vs Deny rules: if we only support FQDN Allow rules, then we can say "the FQDN rule only works if you make a proper DNS request", and we can assume users will do that because they want their traffic to be allowed. But if we support FQDN Deny rules, we can't do that, because we have to assume that attackers will "cheat" in whatever way they can to get around the limits of the implementation.)
- Do we expect/require the rule to work correctly even when the domain name in question has rapidly-changing IPs and does not necessarily return all of its IPs in response to any single DNS query? (eg, if it returns different IPs to pods in one zone than it does to pods in another zone).

Reviewer 3: > If we support both Allow and Deny rules, what happens if, for example, someone allows connections to example.com but denies connections to example.net (which has the same IP).

There is a pre-defined priority for such cases, except for one case: if the rules are listed in the same `to/from` list, which doesn't seem to have the priority semantics (yet?). But generally speaking, this problem also applies to the same pod being selected with 2 different labels, and having allow and deny rules for those. So I would say it is a valid, but probably separate problem.

> do we expect the FQDN rule to work correctly even if the pod got the IP address via some means unknown to us?

I am not sure that is possible, and would say no. As I see it, this feature is just a way to translate a dns name to a set of ips (would like to hear other opinions on this though). In that case, it makes sense to require a dns request to be sent for a dns name resolution. I imagine implementations may have more restrictions on how exactly this should happen, but I don't think we can come up with the common requirements.

> Do we expect/require the rule to work correctly even when the domain name in question has rapidly-changing IPs and does not necessarily return all of its IPs in response to any single DNS query? (eg, if it returns different IPs to pods in one zone than it does to pods in another zone).

Can this also be an implementation detail? I guess it doesn't change anything from the API point of view, and there are probably more corner-cases, e.g. replies with TTL=0 and so on.

Reviewer: > > If we support both Allow and Deny rules, what happens if, for example, someone allows connections to example.com but denies connections to example.net (which has the same IP).
> 
> There is a pre-defined priority for such cases

No, priority only matters for rules that match the same things.

If I say "allow to 1.2.3.4 with priority 100" and "deny to 5.6.7.8 with priority 50" (and nothing else), then the priorities are irrelevant; all traffic to 1.2.3.4 is allowed, and all traffic to 5.6.7.8 is denied.

Likewise if I say "allow to redhat.com with priority 100" and "deny to google.com with priority 50"; all connections to redhat.com are allowed and all connections to google.com are denied.

So if I say "allow to example.com with priority 100" and "deny to example.net with priority 50", the expectation is that all traffic to example.com is allowed, and all traffic to example.net is denied, but what will _actually_ happen is that all traffic to both example.com and example.net will be denied, because the rules effectively become "allow to 93.184.216.34 with priority 100" and "deny to 93.184.216.34 with priority 50", and so the lower-precedence allow rule is overridden by the higher-precedence deny rule.

(There is really only one sane thing we can do here, which is to say "don't do that then", and a few of the implementations linked below are explicit about that. e.g., Azure notes that "FQDN filtering in application rules for HTTP/S ... can discern between two FQDNs that are resolved to the same IP address. This is not the case with FQDN filtering in network rules.")

> > do we expect the FQDN rule to work correctly even if the pod got the IP address via some means unknown to us?
> 
> I am not sure that is possible, and would say no.

In the case of not-rapidly-changing IPs, and non-wildcard, the policy implementation can resolve all the FQDNs to IPs ahead of time; this is how the existing OpenShift EgressFirewall feature works.

My point though was more that if we say "no" here, then that argues that we should _not_ support FQDN Deny rules, since I can't see any way to make them actually secure in that scenario.

> > Do we expect/require the rule to work correctly even when the domain name in question has rapidly-changing IPs
>
> Can this also be an implementation detail?

No, I don't think so; as I noted [below](https://github.com/kubernetes-sigs/network-policy-api/pull/134#discussion_r1331750773), both of the two implementations that use proactive/async DNS resolution also include warnings that they don't work well for rapidly-changing IPs, whereas none of the implementations that use DNS snooping have such a warning. So that implies that if we want to support the rapidly-changing IP use case, then are imposing certain technology choices on implementors.

Reviewer: Well, OK, we _could_ leave "rapidly-changing IPs" as an implementation detail, and say that at the ANP API level, it's undefined whether they will work. But I would consider that very unsatisfying...

Reviewer 3: >No, priority only matters for rules that match the same things.

Exactly, that is why we have priorities: to resolve "conflicts" for the same matches. If we could guarantee the same thing is never selected by different rules, we wouldn't need priorities at all. Your example kinda shows how priorities make these overlapping rules behave somewhat predictably. The same will happen if you "allow to pod selector label1=A with priority 100" and "deny to pod selector label2=B with priority 50", and then have a pod with labels `label1=A, label2=B`. So I think adding extra docs for dns examples, and for other similar examples will be useful here.

What I wanted to stress, is that there is a way to create rules saying `"allow to example.com with priority 100" and "deny to example.net with priority 100"` by just having these rules in the same ANP peers list (and the same can happen with the existing pod selectors). In that case we will have some kind of an undefined behaviour, because there is no strict prioritization between these rules, and the same thing will be allowed and denied with the same priority.

> Well, OK, we could leave "rapidly-changing IPs" as an implementation detail, and say that at the ANP API level, it's undefined whether they will work. But I would consider that very unsatisfying... 

It is certainly nice to have pre-defined behaviour, but it may be very difficult. For example, one plugin may support rapidly-changing ips at every 5 seconds rate, and another one at every millisecond rate. Defining such details will cause "imposing certain technology choices on implementors" as you said, which doesn't sounds like a flexible API.

Author: I don't know that "imposing certain technology choices on implementors" is a bad thing, so long as our decision is "reasonable" and backed by legitimate use-cases. This will be an optional spec of the API, so I hope it should be ok to consulting a bunch of providers and pick a compromise.

For example, the wild-card specification makes using a polling-based solution pretty much unworkable. But there is the fact that lots of providers have made it work, and  we see user-stories that ask for this -- both result in us considering it for the spec.

* In the rapidly-changing IP case, maybe it makes sense to specify a minimum TTL we expect providers to handle correctly (maybe 5 minutes). Providers can certainly do better if their implementation is capable.
* In the allow/deny case, I haven't heard of any user stories for deny that feel super rigorous to me. I'm of the opinion that deny shouldn't be part of the spec. If providers have deny implementations, I don't think keeping it out of the spec is a huge deal (users can still rely on the implementation-specific CRs if they really need it). Happy to reverse that if we find compelling use-cases.

* Provide a selector to specify egress peers using a Fully Qualified Domain Name
  (for example `kubernetes.io`).
* Support a restricted set of regex matching capabilities when specifying FQDNs. 

## Non-Goals

* This enhancement does not include a FQDN selector for allowing ingress
  traffic.
* This enhancement does not include any L7 matching or filtering capabilities,
  like matching HTTP traffic or URL paths.
    * This selector should not control what DNS records are resolvable from a
      particular workload.
* This enhancement does not provide a mechanism for selecting in-cluster
  endpoints using FQDNs. This is explicitly disallowed by the spec.

>     * To select Pods, Nodes, API Server, AdminNetworkPolicy has more first party

Reviewer 4: which are NPEPs by themselves?

Author: Partially. Pods is core spec, Nodes is covered by #126. API Server is an interesting case, and so far we don't have a good selector for it yet.

      selector with better UX.
* This enhancement does not specify the details of how traffic is routed to the
  specified destination. For example, it does not prescribe details around NAT
  or egress gateways.
* This enhancement does not require a particular mechanism for securing DNS
  resolution (e.g. DNSSEC or DNS-over-TLS).
    * TODO: Is this up to implementations? Maybe an optional enum describing

>       what security mechanims to use and providers can chose whther to support

Reviewer: ```suggestion
      what security mechanisms to use and providers can chose whether to support
```

>       what security mechanims to use and providers can chose whther to support

Reviewer: It doesn't require "particular" mechanisms for securing DNS? Does it require _any_ mechanisms?

Author: clarified.

      or not.

## Introduction

> FQDN-based egress controls are a common enterprise security practice.

Reviewer: [citation required]

People _want_ this functionality, but do they already have it outside of Kubernetes? You certainly can't do it with iptables/nftables. (You can say "`iptables -d kubernetes.io -j DROP`" but that means "resolve the IP addresses of `kubernetes.io` right now and create one static rule for each IP address it returns".)

Do external firewall products implement this? If so, how _exactly_ do they implement it, since if that's what users expect, then we should just do the same thing.

Author: I think there's a ton of prior art both inside and outside Kubernetes on some flavor of FQDN firewall.

- https://learn.microsoft.com/en-us/azure/firewall/fqdn-filtering-network-rules
- https://cloud.google.com/firewall/docs/firewall-policies-rule-details#domain-names-for-firewall
- https://docs.vmware.com/en/VMware-NSX/4.1/administration/GUID-63262728-CA72-47D2-8E4F-16617B63A9A4.html
- https://techdocs.f5.com/kb/en-us/products/big-ip-afm/manuals/product/network-firewall-policies-implementations-13-1-0/3.html
- https://docs.tigera.io/calico-enterprise/latest/network-policy/domain-based-policy
- https://docs.cilium.io/en/stable/security/policy/language/#dns-based
- https://docs.openshift.com/container-platform/4.13/networking/openshift_sdn/configuring-egress-firewall.html

As far as matching user expectation, I'm mostly focused on what existing K8s CNIs are providing as that's the real target demographic here. I will do some reading and definitely the goal of the NPEP is to find the least common denominator behaviors. I'm curious if you have a particular behavior you're wary of.

Reviewer: OpenShift's current implementation (based on proactively resolving hostnames in FQDN rules and re-resolving when the TTL expires) has turned out to not work well at all for hosts behind load balancers/CDNs (which seem to be a lot of the hosts that people want to make rules involving).

We're probably moving to an approach where we intercept client DNS requests and update the rules based on the actual IPs returned to the client, but this isn't implemented yet, and it's unknown if it will actually work better.

Reviewer: OK, it looks like:

||Azure|GCP|VMware|F5|Calico|Cilium|OpenShift (current)|OpenShift (future)|
|-|-|-|-|-|-|-|-|-|
|Implementation| DNS snooping | async DNS | DNS snooping | DNS snooping? | DNS snooping | DNS snooping | async DNS | DNS snooping |
|Wildcards?| :heavy_check_mark: | :x: | :heavy_check_mark: | :x: | :heavy_check_mark: | :heavy_check_mark: | :x: | :heavy_check_mark: |
|Egress rules? | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
Ingress rules? | :x: | :heavy_check_mark:| :x:? |:heavy_check_mark: | :x: | :x: | :x: | :x: |
Allow rules? | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
Deny rules? | :heavy_check_mark:? | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :x:? | :x: | :heavy_check_mark: | :x:? |

So...

- There seems to be a strong consensus for DNS snooping, wildcards, and "allow egress" rules.
- Most systems have "deny egress" rules.
- Most systems don't have ingress FQDN rules.
- GCP and (current) OpenShift, the only two that (definitely) use async DNS rather than DNS snooping, both explicitly warn in their docs that they don't work well with frequently-changing DNS names. None of the DNS snooping systems have such warnings.
  - GCP: "For highly dynamic domain names, there might be a discrepancy between the resolved IP addresses in the local cache and the resolved IP addresses in the firewall policy rule. In such a scenario, the egress firewall rule works reliably only when the VM uses the resolved IP addresses configured in the firewall policy rules." ... "Don't use ingress rules that use FQDN objects if the domain name resolution results are highly variable or domain name resolution uses a form of DNS-based load balancing."
  - OpenShift: "Because the egress firewall controller and pods asynchronously poll the same local name server, the pod might obtain the updated IP address before the egress controller does, which causes a race condition. Due to this current limitation, domain name usage in EgressNetworkPolicy objects is only recommended for domains with infrequent IP address changes."

Reviewer: yup, go ahead

Administrators often prefer to write security policies using DNS names such as
“www.kubernetes.io” instead of capturing all the IP addresses the DNS name might
resolve to. Keeping up with changing IP addresses is a maintenance burden, and
hampers the readability of the network policies.

> ## User Stories

Reviewer 2: Can you please add some explicit "real world" examples to these?

Reviewer 2: Hrm also these two use cases are pretty similar... 

For now :) this is an NPEP aiming to add FQDN selector functionality to ANP (I know it obvious for use here but we should probably add a note somewhere as well).... ANP already provides means of selecting the "cluster as a whole" OR "certain workloads in the clsuter"

I think we can narrow these user stories down to a single one and provide some explicit examples for both scenarios. 

something like "As a cluster admin, I want to apply explicit rules to my cluster which allow users to specify FQDNs as valid peers."  And then add some examples of doing so.

Author: Added a couple more details.

> ## User Stories

Reviewer: I think an important user story here is "As a cluster admin, I want to block all cluster egress traffic by default, and require namespace admins to create NetworkPolicies explicitly allowing egress to the domains they need to talk to."

We aren't solving that user story _right now_, but we want to make sure we remember it in the NPv2 NPEP. (And we want to make sure that we don't design ANP FQDN egress in a way that would make DNP/NPv2 FQDN egress not work.)

Author: added.

> * As a cluster admin, I want to allow certain workloads to send traffic to a

Reviewer 2: nit:
```suggestion
* As a cluster admin, I want to allow certain workloads to send traffic to an external
```

  service specified by a well-known domain name (e.g. on-prem logging service)
* As a cluster admin, I want to allow the cluster to communitcate with services

>   provided by a Cloud provider by allowing their whole domain.

Reviewer 3: should we include wildcarded dns names here too? (looks like a popular option anyway)

## API

TODO

## Alternatives

### IP Block Selector

IP blocks are an important tool for specifying Network Policies. However, they
do not address all user needs and have a few short-comings when compared to FQDN
selectors:

* IP-based selectors can become verbose if a single logical service has numerous
  IPs backing it.
* IP-based selectors pose an ongoing maintanance burden for administrators, who
  need to be aware of changing IPs.
* IP-based selectors can result in policies that are difficult to read and
  audit.

### L7 Policy

Another alternative is to provide a true L7 selector, similar to the policies
provided by Service Mesh providers. While L7 selectors can offer more
expressibility, they often come trade-offs that are not suitable for all users:

* L7 selectors necessarily support a select set of protocols. Customers may be
  using a custom protocol for application-level communication, but still want
  the ability to specify endpoints using DNS.
* L7 selectors often require proxies to perform deep packet inspection and
  enforce the policies. These proxies can introduce un-desireable latencies in
  the datapath of applications.

> ## References

Reviewer 3: maybe adding links to the existing implementations of FQDN-based selector would be useful?
